### DISCUSSION TITLE
Github Issue Labels

### DISCUSSION THREAD

**Discussion by avidrucker** [athensresearch/athens#1055] [Operations]
These are the labels we use and some guidelines for their use. The goal of this post is to be a discussion for modifying labels and labeling processes. A more formalized version of our labeling system will be located in the handbook.

#### Priority (based on [Flutter](https://github.com/flutter/flutter/wiki/Issue-hygiene#priorities))
- P0 - "The world is on fire."
    - Should be = 0.
    - Data loss issue.
    - Athens is frozen on startup or on use — not being able to input is a kind of data loss.
- P1 - "A customer is on fire."
    - Should have less than 1 or 2
- P2 - "A customer is about to be on fire."
- P3
    - The highest priority level a bug can have that isn't affecting a top-tier customer or breaking the build.
    - Should be working on unless assignee is dealing with P0-P2 bug.
- P4
    - Default level for bugs. May not be fixed for a long time. Typically will migrate to P3 before we work on them unless part of [Roadmap](https://github.com/flutter/flutter/wiki/Roadmap).
- P5
    - Default level for new feature requests. Valid but not important. Unlikely to work on such issues unless part of a long-term strategic plan.
- P6
    - Unlikely to ever be worked on. Use 👍 to promote to P5 or higher based on demand.

#### Type
- t/bug

#### Operating System
- os/mac
- os/linux
- os/windows

### tasks
- [ ] Create Issue label descriptions #1208 

### questions
- should issue labeling be open to all with a disclaimer that some should be exclusively used by core team? or maybe a core team label to get their attention?

### notes
- from @shanberg 
  - sometimes it's nice for the palette to reflect the category of label, like all priority labels are reddish, domain labels are greenish, etc
  - or for the label color to reflect the severity/importance of the label, like P0 is strong red, P6 is weak red
- Conversation from Discord:
```
@avidrucker : How may I get the ability to assign labels to our GitHub issues? @Johnny @Jeff Tang 🏛
@tangjeff0 : what is your goal @avidrucker
avidrucker: Label new issues, as well as label issues I am working on directly or supporting
Primary goal: Increase interaction & contribution
Marking issues as needing attention, beginner friendly, increased documentation, etc.
Secondary goal: Hopefully to reduce the dependency on Discord as the driver for interaction
```

**Comment by jefftangx** [comment]
Please create requests on this discussion, and not the issues themselves, e.g. here https://github.com/athensresearch/athens/issues/997
![image](https://user-images.githubusercontent.com/8952138/116283640-f2e87c00-a740-11eb-9497-603f92bad3be.png)

We can batch the requests and actions, which will reduce the noise on the issues. Thank you for looking into this @avidrucker !

**Comment by avidrucker** [comment]
Proposed labels:  
- [ ] ux: issues directly affecting the user experience that aren't necessarily features or bugs
- [ ] frontend: components, hiccup, CSS, etc.
- [ ] backend: databases, parsers, APIs, etc,
- [ ] brainstorm: for issues needing ideation
- [ ] feedback-needed: for issues that need a weigh-in or input from others
- [ ] discuss: for issues currently in discussion (not brainstorming or feedback) (note: rather than issues, use "discussions" to discuss, where the issue can link to the discussion)
- [ ] paused-issue: issues that are not closed, but are not being worked on currently
- [ ] pair-programming: issues marked to be done by two people simultaneously, recorded or not
- [ ] needs-testing
- [ ] f/test
- [ ] performance

**Reply by jefftangx** [reply]
Do we have any labels that we want to remove?

What are your thoughts on namespacing labels, such as `d/easy` `d/medium` `d/hard`?

**Reply by avidrucker** [reply]
Difficulty is relative (re: "Simple Made Easy"), and not particularly useful IMO.
Rather, I'd encourage scoping by rough time or rough complexity.
(Rough time is the most helpful IMO)

Rough time to complete:
`15 min`, `30 min` `1 hour`, `1 hour+`
The idea here is that people can set expectations, and then assess if they are on target. It's not a one size fits all, just a rough approximate.

Rough complexity:
`very-simple` `simple` `average`  `complex` `very-complex`
This isn't as useful IMO, but it can be helpful to encourage people to go for things that are challenging / multi-faceted to simplify them, versus someone else wanting something atomic, simple, and constrained.

**Reply by avidrucker** [reply]
Namespacing is great IMO, though it's not critical. Nice to have, can live without.

**Reply by avidrucker** [reply]
So I'd advocate to leave in `beginner-friendly` or `simple` for clearly defined, well constrained issues. I'd advocate to remote the other "difficulty" tags (`medium` and `hard`) because they don't consistently add clear objective value for contributors.

Rough time estimate tags on the other hand will meet everyone's need for gaining further context on how much effort/time a task/issue will involve.

Plus: It also helps to see who added the label IMO - for e.g. when I or @shanberg mark a CSS issue with X time, we bring years of CSS experience to the "scoping" process, and if someone is not confident with CSS, they can apply their own "fudge" number/factor to get a rough picture of the time/complexity involved.

**Reply by shanberg** [reply]
I slightly favor complexity over difficulty, because complexity describes the work alone, and not the work divided by a theoretical average developer's skill level.

**Reply by jefftangx** [reply]
I agree that difficulty is not good - will delete `easy` `medium` and `hard`.

I like complexity.

Time estimates only work for simple complexity things.

**Reply by avidrucker** [reply]
To increase constructive engagement with issues, I would like **some** issues to be broken down to the 30 min or smaller time estimate. This will help not only junior devs who have less experience with coding, Clojure, and open source in general, but it will also help more senior folks who want to mentor do so without hand holding.

**Reply by sid597** [reply]
how do we define complexity? I think the complexity of any issue depends on what time it is calculated, we freeze time look at its surroundings, factors, and then decide on the complexity and this assumes one has knowledge of their surroundings. In the case of Athens, people who are aware of the codebase and new changes being made to it are few (Jeff, Alex, Stuart). Keeping this in mind I have a few points :
- Deciding on the complexity of an issue can only be done by current core members, but this creates an issue of time commitment from their side. 
- If we plan to carry on the [bounties as a recruiting tool](https://github.com/athensresearch/athens/discussions/1071) I think all the issues we label with complexity "X" should roughly be similar at all times which I think would not be possible. 

I propose to let the contributor figure out the complexity of an issue and if needed discuss it, in case of using issues for hiring whenever the candidate picks up an issue they can then ask core members for analysis of the issue's complexity, this way upfront time commitment would not be needed and the problem of : _all issues with complexity "x" should be similar_, can be solved fairly.
What do you think @tangjeff0 @shanberg @avidrucker ?

**Comment by avidrucker** [comment]
## Labeling To-Do's

- [ ] mark these with the `ux` label  
    - [ ] https://github.com/athensresearch/athens/issues/1048
    - [ ] https://github.com/athensresearch/athens/issues/1016
    - [ ] https://github.com/athensresearch/athens/issues/566

- [ ] candidates for test verification:  
    - [ ] https://github.com/athensresearch/athens/issues/971
    - [ ] https://github.com/athensresearch/athens/issues/969

- [ ] mark with labels `discuss`, `architecture`, `backend`, `frontend`, `feedback-needed`
    - [ ]  https://github.com/athensresearch/athens/issues/997

- [ ] mark with label `performance`
    - [ ] https://github.com/athensresearch/athens/issues/977

**Reply by jefftangx** [reply]
These look all good to mark! @avidrucker

**Comment by avidrucker** [comment]
@shanberg Is the design label only for design tasks, or for anything design related?  
- [ ] See issue: https://github.com/athensresearch/athens/issues/1041

**Reply by jefftangx** [reply]
I think anything that is design-related. But what is the difference between `ux` and `design`?

**Comment by sid597** [comment]
More proposed labels, inspired by flutter's labels: 
- customer/ :  _Can use this lable to mention if a customer raised the issue_
   - Enterprise
   - user
   - open source sponsor
- depends-on/
   - person: 
   - Other feature:
   - Plugin architecture: _First need plugin architecture to implement this feature_
   - Text editor: _Cannot solve this bug or implement this feature without a new text editor_
   - ..... _All other features that are in the roadmap, so that we can see what current feature requests or bugs depending on the feature in roadmap_
- found in release version/ 
  - ... mention last2/3 latest versions
- Need/
  - opinion
  - review
  - help
  - feedback
  - test
- Platform/
  - Computer/laptop
  - Mobile
- Status/ : _Current status of an issue_
  - assigned
  - open
  - PR raised
  - reviewing
  - brainstorming
  - self-assigned
- Issue/ : _state of issue_
  - fixed: _can be a hack, not solution_
  - solved
  - duplicate
  - timeout: _author not responding_
  - invalid
- tech-debt: _An issue arose because of previously written code_
- proposal
- easy-fix --> good first issue
- roadmap/ : _feature in `nth` month/year of roadmap_
  - ... values for `n`
- Not in the roadmap   
- Remote: _related to remote db_
- Has reproducible steps
- Feature request by or agreed upon(can be via emoji or comment) by a core team member: _This increases the priority of the request_
- help popup: _A help popup would be good for this issue_
-  User confused by behaviour: _user is confused how to use some behaviour and if it is the right way to do this_

**Reply by sid597** [reply]
- assigned for triage: Sometimes it's not clear what the appropriate labels are for an issue, or you suspect it may be a duplicate but you lack enough domain knowledge to know for sure. If you know someone who has the expertise to triage the issue, you can assign the issue to them and add the label assigned for triage, so the person knows that they are not expected to fix it, but to triage it.

**Reply by sid597** [reply]
- doc label: If the bug report is a question, then it probably belongs in Stack Overflow or on our #help channel or some other forum for getting help. However, if it appears that the reporter did try to read our documentation to find the answer, and failed, or, if you look in our documentation and find it is inadequate here, then please consider it a documentation bug (and update the summary accordingly).

**Reply by sid597** [reply]
If it seems to affect a lot of people but doesn't quite rise to the level of a release blocker, add the `customer: crowd` label (which means it affects many people) and then either the P1 label (if it's blocking people) or the P2 label (if it's not blocking people but is nonetheless serious). This will flag the bug for consideration during weekly triage (see the second section of this document).

**Reply by sid597** [reply]
Can also divide `feature request` label to :
- feature/ 
  - request
  - extend: _Feature is already there and issue is asking to add more to it for example #859_

**Comment by jsmorabito** [comment]
we might like to have a group of 'first good issue' labels that specify for whom the issue is good for (front-end, backend, design, etc)

**Reply by jsmorabito** [reply]
looks like by the way github leverages it we'll need to simply use the 'good first issue' label then further labels after that can help the potential contributor identify if its an issue within the domain theyre learning. 
![image](https://user-images.githubusercontent.com/46250921/121590977-f8ef9f00-ca06-11eb-9047-cfdd971dbb40.png)
